### PR TITLE
Implement the `foreach` loop

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR implements the `foreach` loop as proposed in
https://github.com/chapel-lang/chapel/issues/16404.

Co-developed with @mppf
Supersedes his https://github.com/chapel-lang/chapel/pull/17014

#### Background
`foreach` loops are the user facing way of implementing sequential yet
vectorizable loops. Before this PR, the vectorization support relied on non-user
facing pragmas to be attached to the iterator symbols. With this PR, those
pragmas are no longer necessary if the iterator uses `foreach` loops.

#### Implementation
The implementation is rather straightforward. `foreach` loops are just
`ForLoop`s that are made order independent from the time they are created.
`foreach`-ness of the loop is not stored in the `ForLoop`'s interface.

Similar to what we have today, a `ForLoop` can be made order independent during
resolution if it consists of only a yield and iterates over something that is
also order independent. If that happens to a `for`, after that point it is
indistinguishable from a `foreach` for the rest of the compilation.

Note that we still need the pragma `order independent yielding loops` (and it's
inverse, just for completeness (?) ) to mark iterators that uses non-`for` loops
as order independent. However, with this PR, flags corresponding to those
pragmas are no longer added by the compiler. Instead, the compiler choses
between `for`/`foreach` when writing iterators for `forall` expressions, for
example.

#### Future work
- We need to implement `foreach` intents. (It appears we'll also add `for`
  intents https://github.com/chapel-lang/chapel/issues/17857)

- @bradcray has made a good point that intent implementations for
  `for`/`foreach` and `forall` may end up to be sufficiently different.
  Depending on what we see then, we could think of creating a `ForeachLoop` AST
  node.

- Do we want `foreach` expressions? Currently they fail with a generic syntax
  error.

- We should probably beef up testing, especially for the standard/internal
  module iterators that now use `foreach` to make sure that they are vectorized
  as appropriate. (I can add some more testing in this PR, but maybe not
  exhaustively)

#### Testing

- [x] release/examples with `--fast`
- [x] standard
- [x] gasnet

### CODE REVIEW

File: test/performance/vectorization/vectorPragmas/cForLoopInParIterForeach.chpl [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  // C for loop in standalone with yield should get vector pragma
+  iter myiter(nn: int, nt: int, param tag: iterKind) where tag == iterKind.standalone {
+    coforall i in 0..#nt {
+      const followThis = i*nn..#nn;
+      var idx: followThis.first.type;
+      const first = followThis.first;
+      const str   = followThis.stride;
+      const end   = followThis.last + str;
+      while __primitive("C for loop",
```
Comment (mppf) [inline test/performance/vectorization/vectorPragmas/cForLoopInParIterForeach.chpl:19]: I thought that c for loop required the pragma on the iterator? Is the compiler adjusting this one somewhere?

File: test/performance/vectorization/vectorPragmas/nestedLoopsInFollowerForeach.chpl [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  iter myiter(nn: int, nt: int, param tag: iterKind) where tag == iterKind.leader {
+    coforall i in 0..#nt {
+      yield i*nn..#nn;
+    }
+  }
+
+  // neither loop will get the pragma since this iter cannot be zippered
+  // inline. I this is reasonable since it's pretty tricky to zipper nested
+  // loops together. Not to say it's impossible.
```
Comment (mppf) [inline test/performance/vectorization/vectorPragmas/nestedLoopsInFollowerForeach.chpl:29]: Time to remove this comment?

File: test/performance/vectorization/vectorPragmas/forallInStandaloneForeach.chpl [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module iters {
+  iter myiter(nn: int, nt: int) {
+    foreach i in 0..#nt {
+      foreach j in i*nn..#nn {
+        yield j;
+      }
+    }
+  }
+
+  // coforall loop in standalone should NOT get vector pragma
```
Comment (mppf) [inline test/performance/vectorization/vectorPragmas/forallInStandaloneForeach.chpl:10]: This comment is not making sense to me. Maybe this and the next should just say the forall should get a vector pragma?

File: test/performance/vectorization/vectorPragmas/zipIterInFollowerForeach.chpl [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    foreach i in 0..#nt {
+      foreach j in i*nn..#nn {
+        yield j;
+      }
+    }
+  }
+
+  // zippered loop in standalone with yield should get vector pragma
+  // Note: requires that each iter in the zippered iter gets inlined
+  pragma "order independent yielding loops"
```
Comment (mppf) [inline test/performance/vectorization/vectorPragmas/zipIterInFollowerForeach.chpl:12]: remove this pragma and it should still work the same, right?

File: test/performance/vectorization/vectorPragmas/zipIterInFollowerForeach.chpl [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  iter myiter(nn: int, nt: int, param tag: iterKind) where tag == iterKind.leader {
+    coforall i in 0..#nt {
+      yield i*nn..#nn;
+    }
+  }
+
+  // zippered loop in standalone with yield should get vector pragma
+  // Note: requires that each iter in the zippered iter gets inlined
+  pragma "order independent yielding loops"
```
Comment (mppf) [inline test/performance/vectorization/vectorPragmas/zipIterInFollowerForeach.chpl:29]: ditto

File: test/performance/vectorization/vectorPragmas/zipIterInFollowerNoVectorForeach.chpl [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  iter nonInlinableIter(r: range) {
+    foreach i in r # (r.size - 1) do 
+      yield i;
+    yield r.last;
+  }
+
+  iter myiter(nn: int, nt: int) {
+    foreach i in 0..#nt {
+      foreach j in i*nn..#nn {
```
Comment (mppf) [inline test/performance/vectorization/vectorPragmas/zipIterInFollowerNoVectorForeach.chpl:11]: it seems to me that this test would be better if the serial version is also not inlineable

**Review by mppf — APPROVED** [review]
Great!

**Comment by mppf** [comment]
I forgot to mention in my review but we'll need to make spec updates. I'd recommend creating a separate PR now to do that and linking it from here. Or manage it some other way but it's important not to forget.

File: test/performance/vectorization/vectorPragmas/forallInStandaloneForeach.chpl (hunk 3 above)
Comment (e-kayrakli) [reply test/performance/vectorization/vectorPragmas/forallInStandaloneForeach.chpl:10]: This test is based on `forallInStandalone.chpl` in the same directory, which also has the same comment :)

I think it is referring to the `coforall` and `for` created while lowering the forall inside the standalone. Or at least, that's how I read it.

File: test/performance/vectorization/vectorPragmas/cForLoopInParIterForeach.chpl (hunk 1 above)
Comment (e-kayrakli) [reply test/performance/vectorization/vectorPragmas/cForLoopInParIterForeach.chpl:19]: This is a just-yield loop, so the compiler marks it without the pragma

**Comment by e-kayrakli** [comment]
https://github.com/Cray/chapel-private/issues/2245 refers to need for documentation. I'll do some more testing and merge next week.

**Comment by bradcray** [comment]
I haven't had a chance to look at this PR yet, but was stewing more on what it would take to retire that pragma I asked about yesterday altogether (you mentioned some lingering cases of it, like I think some while loops?).  Even if we don't prioritize it in the short-term, do we have an issue capturing the notion of retiring that pragma, or looking into what would be necessary to do so?  (or maybe it's evident to those who know it better than I do that we will want it for all time...?).

**Comment by e-kayrakli** [comment]
I don't think there's an issue, but the crux of what would be needed is either

- to have order-independent versions of `while` and `do-while` loops, or
- rewrite them somehow using `foreach` in the standard/internal modules

There are also cases where we do `__primitive("c for loop", ...)` which we ideally should be able to get rid of somehow, but I don't know what it'd take.

I can open an issue with more detail about this.

**Pull request merged**